File: go/tracing/publisher.go
package tracing

import (
	"context"
	"log"
	"time"

	"github.com/redis/go-redis/v9"
)

// Publisher publishes trace events to Redis Pub/Sub.
type Publisher struct {
	client  *redis.Client
	channel string
	enabled bool
}

// PublisherConfig contains configuration for the Publisher.
type PublisherConfig struct {
	RedisURL string
	Channel  string
	Enabled  bool
}

// DefaultConfig returns default publisher configuration.
func DefaultConfig() *PublisherConfig {
	return &PublisherConfig{
		RedisURL: "redis://localhost:6379",
		Channel:  "jarvis:traces",
		Enabled:  true,
	}
}

// NewPublisher creates a new trace publisher.
func NewPublisher(cfg *PublisherConfig) (*Publisher, error) {
	if cfg == nil {
		cfg = DefaultConfig()
	}

	if !cfg.Enabled {
		return &Publisher{enabled: false}, nil
	}

	opt, err := redis.ParseURL(cfg.RedisURL)
	if err != nil {
		return nil, err
	}

	client := redis.NewClient(opt)

	// Test connection
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	if err := client.Ping(ctx).Err(); err != nil {
		return nil, err
	}

	log.Printf("TracePublisher connected to Redis: %s", cfg.RedisURL)

	return &Publisher{
		client:  client,
		channel: cfg.Channel,
		enabled: true,
	}, nil
}

// NewPublisherFromClient creates a publisher from existing Redis client.
func NewPublisherFromClient(client *redis.Client, channel string) *Publisher {
	return &Publisher{
		client:  client,
		channel: channel,
		enabled: true,
	}
}

// Close closes the Redis connection.
func (p *Publisher) Close() error {
	if p.client != nil {
		return p.client.Close()
	}
	return nil
}

// Publish sends a trace event to Redis.
func (p *Publisher) Publish(ctx context.Context, event *TraceEvent) error {
	if !p.enabled || p.client == nil {
		return nil
	}

	data, err := event.ToJSON()
	if err != nil {
		return err
	}

	return p.client.Publish(ctx, p.channel, data).Err()
}

// PublishStart publishes a "started" event for an operation.
// Returns the span ID for tracking.
func (p *Publisher) PublishStart(
	ctx context.Context,
	operation string,
	service ServiceName,
	metadata map[string]any,
	tags map[string]string,
) string {
	traceID := GetTraceID(ctx)
	if traceID == "" {
		return ""
	}

	spanID := GetSpanID(ctx)
	if spanID == "" {
		spanID = GenerateSpanID()
	}

	event := NewTraceEvent(traceID, spanID, service, operation, StatusStarted).
		SetParentSpan(GetParentSpanID(ctx))

	for k, v := range metadata {
		event.AddMetadata(k, v)
	}
	for k, v := range tags {
		event.AddTag(k, v)
	}

	if err := p.Publish(ctx, event); err != nil {
		log.Printf("Failed to publish trace start: %v", err)
	}

	return spanID
}

// PublishEnd publishes a completion event.
func (p *Publisher) PublishEnd(
	ctx context.Context,
	operation string,
	service ServiceName,
	spanID string,
	status TraceStatus,
	durationMs int,
	metadata map[string]any,
	tags map[string]string,
	errorMsg string,
) {
	traceID := GetTraceID(ctx)
	if traceID == "" {
		return
	}

	event := NewTraceEvent(traceID, spanID, service, operation, status).
		SetParentSpan(GetParentSpanID(ctx)).
		SetDuration(durationMs)

	if errorMsg != "" {
		event.SetError(errorMsg)
	}

	for k, v := range metadata {
		event.AddMetadata(k, v)
	}
	for k, v := range tags {
		event.AddTag(k, v)
	}

	if err := p.Publish(ctx, event); err != nil {
		log.Printf("Failed to publish trace end: %v", err)
	}
}

// Span represents an active trace span for convenient timing.
type Span struct {
	publisher  *Publisher
	ctx        context.Context
	operation  string
	service    ServiceName
	spanID     string
	startTime  time.Time
	metadata   map[string]any
	tags       map[string]string
	errorMsg   string
	isFinished bool
}

// StartSpan creates and publishes a new span.
func (p *Publisher) StartSpan(ctx context.Context, operation string, service ServiceName) *Span {
	spanID := p.PublishStart(ctx, operation, service, nil, nil)

	return &Span{
		publisher: p,
		ctx:       ctx,
		operation: operation,
		service:   service,
		spanID:    spanID,
		startTime: time.Now(),
		metadata:  make(map[string]any),
		tags:      make(map[string]string),
	}
}

// AddMetadata adds metadata to the span.
func (s *Span) AddMetadata(key string, value any) *Span {
	s.metadata[key] = value
	return s
}

// AddTag adds a tag to the span.
func (s *Span) AddTag(key, value string) *Span {
	s.tags[key] = value
	return s
}

// SetError marks the span as failed.
func (s *Span) SetError(err error) *Span {
	if err != nil {
		s.errorMsg = err.Error()
	}
	return s
}

// Finish completes the span and publishes the end event.
func (s *Span) Finish() {
	if s.isFinished {
		return
	}
	s.isFinished = true

	durationMs := int(time.Since(s.startTime).Milliseconds())
	status := StatusSuccess
	if s.errorMsg != "" {
		status = StatusError
	}

	s.publisher.PublishEnd(
		s.ctx,
		s.operation,
		s.service,
		s.spanID,
		status,
		durationMs,
		s.metadata,
		s.tags,
		s.errorMsg,
	)
}

// FinishWithError completes the span with an error.
func (s *Span) FinishWithError(err error) {
	s.SetError(err)
	s.Finish()
}

File: python/duq_tracing/publisher.py
"""
Trace Publisher
===============

Async Redis publisher for trace events.
Publishes events to Redis Pub/Sub channel for real-time consumption.
"""

import asyncio
import logging
from datetime import datetime
from typing import Any, Optional

from .context import (
    SpanContext,
    generate_span_id,
    get_parent_span_id,
    get_span_id,
    get_trace_id,
)
from .models import Operations, ServiceName, TraceEvent, TraceStatus

logger = logging.getLogger(__name__)


class TracePublisher:
    """
    Async trace event publisher using Redis Pub/Sub.

    Example:
        publisher = TracePublisher(redis_url="redis://localhost:6379")
        await publisher.connect()

        await publisher.publish(TraceEvent(
            trace_id="...",
            span_id="...",
            service=ServiceName.BACKEND,
            operation="llm_call",
            status=TraceStatus.SUCCESS
        ))

        await publisher.close()
    """

    def __init__(
        self,
        redis_url: str = "redis://localhost:6379",
        channel: str = "duq:traces",
        enabled: bool = True,
    ):
        self.redis_url = redis_url
        self.channel = channel
        self.enabled = enabled
        self._redis = None

    async def connect(self) -> None:
        """Connect to Redis."""
        if not self.enabled:
            return

        try:
            import redis.asyncio as redis

            self._redis = redis.from_url(self.redis_url)
            # Test connection
            await self._redis.ping()
            logger.info("TracePublisher connected to Redis: %s", self.redis_url)
        except Exception as e:
            logger.error("Failed to connect to Redis: %s", e)
            self._redis = None

    async def close(self) -> None:
        """Close Redis connection."""
        if self._redis:
            await self._redis.close()
            self._redis = None

    async def publish(self, event: TraceEvent) -> bool:
        """
        Publish a trace event to Redis channel.

        Returns True if published successfully, False otherwise.
        """
        if not self.enabled or not self._redis:
            return False

        try:
            message = event.model_dump_json()
            await self._redis.publish(self.channel, message)
            logger.debug("Published trace event: %s %s", event.operation, event.span_id)
            return True
        except Exception as e:
            logger.error("Failed to publish trace event: %s", e)
            return False

    async def publish_start(
        self,
        operation: str,
        service: ServiceName,
        metadata: Optional[dict[str, Any]] = None,
        tags: Optional[dict[str, str]] = None,
    ) -> Optional[str]:
        """
        Publish a "started" event for an operation.

        Returns the span_id for tracking, or None if publishing failed.
        Uses current trace context.
        """
        trace_id = get_trace_id()
        if not trace_id:
            logger.warning("No trace_id in context for %s", operation)
            return None

        span_id = get_span_id() or generate_span_id()

        event = TraceEvent(
            trace_id=trace_id,
            span_id=span_id,
            parent_span_id=get_parent_span_id(),
            service=service,
            operation=operation,
            status=TraceStatus.STARTED,
            metadata=metadata or {},
            tags=tags or {},
        )

        await self.publish(event)
        return span_id

    async def publish_end(
        self,
        operation: str,
        service: ServiceName,
        span_id: str,
        status: TraceStatus,
        duration_ms: int,
        metadata: Optional[dict[str, Any]] = None,
        tags: Optional[dict[str, str]] = None,
        error: Optional[str] = None,
    ) -> bool:
        """
        Publish a completion event (success or error).
        """
        trace_id = get_trace_id()
        if not trace_id:
            return False

        event = TraceEvent(
            trace_id=trace_id,
            span_id=span_id,
            parent_span_id=get_parent_span_id(),
            service=service,
            operation=operation,
            status=status,
            duration_ms=duration_ms,
            metadata=metadata or {},
            tags=tags or {},
            error=error,
        )

        return await self.publish(event)


class TracedOperation:
    """
    Async context manager for tracing an operation.

    Automatically publishes start/end events and tracks duration.

    Example:
        async with TracedOperation(
            publisher, "llm_call", ServiceName.BACKEND,
            metadata={"model": "claude-sonnet"}
        ) as span:
            result = await call_llm()
            span.add_metadata("tokens", 1500)
    """

    def __init__(
        self,
        publisher: TracePublisher,
        operation: str,
        service: ServiceName,
        metadata: Optional[dict[str, Any]] = None,
        tags: Optional[dict[str, str]] = None,
    ):
        self.publisher = publisher
        self.operation = operation
        self.service = service
        self.metadata = metadata or {}
        self.tags = tags or {}
        self.span_context = SpanContext()
        self.span_id: Optional[str] = None
        self.start_time: Optional[float] = None
        self.error: Optional[str] = None

    def add_metadata(self, key: str, value: Any) -> None:
        """Add metadata during operation execution."""
        self.metadata[key] = value

    def add_tag(self, key: str, value: str) -> None:
        """Add tag during operation execution."""
        self.tags[key] = value

    def set_error(self, error: str) -> None:
        """Mark operation as failed with error message."""
        self.error = error

    async def __aenter__(self) -> "TracedOperation":
        import time

        # Enter span context
        self.span_id = self.span_context.__enter__()
        self.start_time = time.perf_counter()

        # Publish start event
        await self.publisher.publish_start(
            self.operation,
            self.service,
            metadata=self.metadata.copy(),
            tags=self.tags.copy(),
        )

        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        import time

        # Calculate duration
        duration_ms = int((time.perf_counter() - self.start_time) * 1000)

        # Determine status
        if exc_val is not None:
            status = TraceStatus.ERROR
            self.error = str(exc_val)
        elif self.error:
            status = TraceStatus.ERROR
        else:
            status = TraceStatus.SUCCESS

        # Publish end event
        await self.publisher.publish_end(
            self.operation,
            self.service,
            self.span_id,
            status,
            duration_ms,
            metadata=self.metadata,
            tags=self.tags,
            error=self.error,
        )

        # Exit span context
        self.span_context.__exit__(exc_type, exc_val, exc_tb)

        return False  # Don't suppress exceptions


# Global publisher instance (set during app initialization)
_publisher: Optional[TracePublisher] = None


def get_publisher() -> Optional[TracePublisher]:
    """Get the global trace publisher."""
    return _publisher


def set_publisher(publisher: TracePublisher) -> None:
    """Set the global trace publisher."""
    global _publisher
    _publisher = publisher
